File: DataStructures/UnionFind.cpp
#include "UnionFind.h"

using namespace std;

UnionFind::UnionFind(const int& size) :data(size, -1) {}

const int UnionFind::get_root(const int& x) {
	return this->data[x] < 0 ? x : this->data[x] = this->get_root(this->data[x]);
}

void UnionFind::union_set(int x, int y) {
	x = this->get_root(x);
	y = this->get_root(y);

	if (x != y) {
		if (this->data[y] < data[x]) {
			swap(x, y);
		}
		this->data[x] += this->data[y];
		this->data[y] = x;
	}
}

const bool UnionFind::is_same(const int& x, const int& y) {
	return this->get_root(x) == this->get_root(y);
}

const int UnionFind::get_size(const int& x) {
	return -this->data[this->get_root(x)];
}

File: DataStructures/UnionFind.py
class UnionFind:
    def __init__(self, size):
        self.data = [-1]*size

    def get_root(self, x):
        if self.data[x] < 0: return x
        else: self.data[x] = self.get_root(self.data[x])
        return self.data[x]

    def unite_set(self, x, y):
        x = self.get_root(x)
        y = self.get_root(y)
        if x != y:
            if self.data[y] < self.data[x]:
                self.data[x], self.data[y] = self.data[x], self.data[y]
            self.data[x] += self.data[y]
            self.data[y] = x

    def is_same(self, x, y):
        return self.get_root(x) == self.get_root(y)

    def get_size(self, x):
        return -self.data[self.get_root(x)]
